src/charm.py

```diff
@@ -390,16 +390,22 @@ def _get_redis_relation_data(self) -> typing.Tuple[str, int]:
         Raises:
             MissingRedisRelationDataError if the some of redis relation data is malformed/missing
         """
-        relation_data = self.redis.relation_data
-        if not relation_data:
+        relation = self.model.get_relation(self.redis.relation_name)
+        if not relation:
             raise MissingRedisRelationDataError("No redis relation data")
+        relation_app_data = relation.data[relation.app]
+        relation_unit_data = self.redis.relation_data
 
         try:
-            redis_hostname = str(relation_data["hostname"])
-            redis_port = int(relation_data["port"])
+            redis_hostname = str(
```

Comment:
As an alternative:

`redis_hostname = str(relation_app_data.get("leader-host", relation_unit_data["hostname"]))`

All the catching of the exceptions is a bit strange, but can be refactored in the future as it is not what this change did.

Reply:
Thats even cleaner! Thank you :D. I will remember this if I do a refactor pr in the future.